Provide ExtensionDtype.construct_from_string by default

datapythonista opened the pull request:
- [ ] closes #xxxx
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

I think it makes sense to provide a standard `construct_from_string` by default, instead of forcing subclasses of `ExtensionDtype` to implement it.

This way we can define a simple dtype with:

```python
class SimpleDtype(pandas.core.dtypes.dtypes.ExtensionDtype):
    name = 'simple'

    @property
    def type(self):
        return object
```

instead of:
```python
class SimpleDtype(pandas.core.dtypes.dtypes.ExtensionDtype):
    name = 'simple'

    @property
    def type(self):
        return object

    @classmethod
    def construct_from_string(cls, string):
        if string != cls.name:
            raise TypeError("Cannot construct a '{}' from '{}'".format(
                cls.__name__, string))
        return cls()
```

CC: @TomAugspurger

datapythonista committed:
Provide Extension.Dtype.construct_from_string by default

jorisvandenbossche reviewed (approved):
Seems a good idea to me.

I think the example in the docstring can then be removed? As the example is then the actual implementation. 
Or update it to say that the default implementation is adequate in case  that the dtype can be constructed without any arguments.

TomAugspurger commented:
Seems reasonable. I'm not sure what the motivation was for not doing that in the first place. Possibly that we internally call `construct_from_string()` inside `try / except TypeError` blocks in a few places, which would potentially mask the issue? But I think our tests should be adequate.

datapythonista committed:
Merge remote-tracking branch 'upstream/master' into construct_from_string

datapythonista committed:
Improving construct_from_string docstring, and remove unnecessary overwrite of the method

pep8speaks commented:
Hello @datapythonista! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/dtypes/base.py`](https://github.com/pandas-dev/pandas/blob/661d5225583514281bb4ff2692a3e17bd1378636/pandas/core/dtypes/base.py):

> [Line 210:48](https://github.com/pandas-dev/pandas/blob/661d5225583514281bb4ff2692a3e17bd1378636/pandas/core/dtypes/base.py#L210): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\['
> [Line 210:66](https://github.com/pandas-dev/pandas/blob/661d5225583514281bb4ff2692a3e17bd1378636/pandas/core/dtypes/base.py#L210): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\]'

datapythonista committed:
Ignoring flake8 false positive

datapythonista commented:
Thanks for the feedback. I fixed the docstrings, improving the example to show how to overwrite the method, and remove the method of the category dtype, since now was the same as the parent class.

I think in a follow up PR we could make a generic `construct_from_string` method that extracts the arguments from a string like `datetime64[ns, UTC]` and returns an instance of the class passing the list of everything in the squared brackets as `*args`. This way we shouldn't have to overwrite `construct_from_string` for any dtype. I guess I'm not missing anything.

jreback commented on pandas/core/dtypes/base.py:
```diff
@@ -174,15 +174,26 @@ def construct_array_type(cls):
     @classmethod
     def construct_from_string(cls, string):
```
can you add types

jreback reviewed (changes requested):
@TomAugspurger IIRC we explicitly turned off the parsing of datetime w/tz dtypes (in the Dtype constructor).

@datapythonista thus I don't think this is an instructive example here (no objection to moving the integer usecase here though);

there is also a usage in intervals I think?

this would need a test for existence of this method as well

Commit:
Merge remote-tracking branch 'upstream/master' into construct_from_string

Commit:
Addressing CI and PR comments: implementing tests, fixing docstring error and removing CategoricalDtype.construct_from_string

Commit:
Adding type annotations

jorisvandenbossche commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -51,6 +51,36 @@ def test_pickle(self):
         assert result == self.dtype
 
 
+class TestBaseDtype:
```
Such tests are already in the base extension tests (tests/extension/base/dtype), or can be expanded there

datapythonista replied:
I had the feeling that those were only base tests for implementing tests for custom extensions, couldn't find a place where I thought it'd make sense to have the abstract class test. Where exactly would you have them?

jorisvandenbossche replied:
>  those were only base tests for implementing tests for custom extensions

Yes, but we have custom extensions ourselves, so we use those tests ourselves (and in addition we also have a few extra test EAs). 

For example in `tests/extension/decimal/array.py` you can remove the custom implementation and let it inherit, and then it is already tested in that way.

> where I thought it'd make sense to have the abstract class test

What do you mean with abstract class test?

datapythonista replied:
I don't think I understand what you want to test.

@jreback wanted a test for the existence of the method `construct_from_string`, which makes sense to me. Ideally, we'd already have a test, testing that the method raises an exception when called directly from `ExtensionDtype` (what I called the abstract class before). Then I'd just replace that by checking that now it's implemented, and it can construct an instance given the string of the name of the dtype.

I saw that in `extension/base/dtype.py` we already have a test for `construct_from_string`, but that's meant for custom extensions (ours or from a third-party) to test that their implementation of `construct_from_string` follows the expected behavior. The way of using that test is to create a subclass of `BaseDtypeTests`.

I don't think we want to test that the method is implemented in the `ExtensionDtype` (the base/abstract class) by checking a subclass of it. So, I see two options:

- Create a subclass of `BaseDtypeTests` defining that the `dtype` to test is not a subclass but `ExtensionDtype` itself (it will fail for the abstract methods that still exist in `ExtensionDtype).
- Create an independent test for `ExtensionDtype` and not their subclasses, that doesn't subclass `BaseDtypeTests` (this is what I've done here, not testing only `construct_from_string` but all the methods that I could tests, since there weren't tests for the the "default"/base implementations).

Am I missing something here? Sorry I don't understand exactly what you propose.

jreback replied:
see @jorisvandenbossche comments on where tests should go, we have a very general framework for these already.

jorisvandenbossche commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -51,6 +51,36 @@ def test_pickle(self):
+        assert isinstance(dtype_instance, self.dtype.__class__)
+        with pytest.raises(TypeError, match="Cannot construct a 'DummyDtype' "
+                                            "from 'another_type'"):
+            self.dtype.__class__.construct_from_string('another_type')
```
I think this would be a useful test to add to the base extension tests, and the `test_eq` as well. The defaults tests are maybe harder though.

Commit:
Merge remote-tracking branch 'upstream/master' into construct_from_string

jreback commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -51,6 +51,36 @@ def test_pickle(self):
 
 
+class TestBaseDtype:
+    def setup_method(self):
```
we really don't want to use setup_method generally at all.

Commit:
Merge remote-tracking branch 'upstream/master' into construct_from_string

Commit:
Moving tests to pandas/tests/extension/

Commit:
Fixing test (error message)

datapythonista commented:
@jorisvandenbossche @jreback I think the tests are in the location you meant now. Let me know if you have any further comment.

jorisvandenbossche commented:
Thanks! That's what I was thinking about. You are happy with that as well?
 
I think there are still some custom implementation of `construct_from_string` (that are now identical as the parent one) in the test arrays that can be removed (in decimal, json, arrow/bool)

datapythonista commented:
I think the tests now are testing something slightly different as before. They're testing that new extension dtypes satisfy the implementation of the base class. Before I was testing the base class itself. I think both are reasonable, I'm happy this way too, but I think it also made sense what I implemented before.

I deleted the two `construct_from_string` methods that were repeating the new base class behavior, there are no more to delete.

Thanks for the review, let me know if there is anything else that should be changed, I think it should be ready now.

jreback commented:
thanks @datapythonista

TomAugspurger commented:
This apparently caused some performance issues: http://pandas.pydata.org/speed/pandas/index.html#sparse.SparseDataFrameConstructor.time_from_scipy?commits=5a724b5cd796a6ede3cb95b8687eaf561e9d57b2

Will open a proper issue later.

jorisvandenbossche commented:
It seems a bit strange that this PR is the cause of that regression, as SparseDtype has its own implementation of `construct_from_dtype`.

jorisvandenbossche commented:
Quick profiling of the specific benchmark points out the culprit: in the dtype checking code if doing a try/except it formats the full series in the error message.

TomAugspurger commented:
Ah, thanks for digging into it!

It's unfortunate that such a minor change can have such broad consequences on performance. I assume this is from something like `pandas_dtype(thing)`, where `thing` is an array-like rather than a dtype? If we were more careful about not passing data to `pandas_dtype`, we would have been OK?

jorisvandenbossche commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -440,19 +440,6 @@ def construct_array_type(cls):
-            else:
-                raise TypeError("cannot construct a CategoricalDtype")
-        except AttributeError:
-            pass
```
Does anybody know why we had this "except AttributeError: pass" here?

jorisvandenbossche commented:
We have had similar performance regressions before. In general we need to be careful with what we print in error messages, but this is of course easy to overlook, as we did in this case. But I would say: that's what we have benchmarks for, and they worked nicely to catch it here.

jorisvandenbossche commented:
I did https://github.com/pandas-dev/pandas/pull/26776 with a possible fix. 

I also don't think it is that an important fix, as there are some strange things going on in `SparseDataFrame._init_spmatrix`, where they pass a dict of {colname: SparseSeries} to `Index.difference` method (while it should pass something like dict.values() I think), and so it is the conversion of this dict to an Index that is taking a lot of time (in there it is checking if this object is categorical, leading to the error message, as dicts are not special cased there, and leading to printing a dict of 1000 serieses). 
Given that this is all deprecated, I am not going to put effort in cleaning that up.
